File: src/Domain/Entities/DoctorAggregate/Doctor.cs
using System.Collections.Generic;
using System.Linq;
using DocHelper.Domain.Interfaces;

namespace DocHelper.Domain.Entities.DoctorAggregate
{
    public class Doctor : BaseEntity, IAggregateRoot
    {
        public string FirstName { get; private set; }
        public string LastName { get; private set; }
        public string MiddleName { get; private set; }
        public string Titles { get; private set; }
        public int WorkExperience { get; private set; }
        public string Description { get; private set; }
        public string Photo { get; private set; }

        public Stats Stats { get; private set; }

        public ICollection<Information> Informations { get; private set; }

        public ICollection<DoctorSpecialty> Specialties { get; private set; }

        private Doctor()
        {
        }

        public Doctor(string firstName, string lastName, string middleName, string titles, int workExperience,
            string description, string photo)
        {
            FirstName = firstName;
            LastName = lastName;
            MiddleName = middleName;
            Titles = titles;
            WorkExperience = workExperience;
            Description = description;
            Photo = photo;
        }

        public Doctor(string firstName, string lastName, string middleName, string titles, int workExperience,
            string description, string photo, Stats stats, List<Information> informations,
            List<DoctorSpecialty> specialties)
        {
            FirstName = firstName;
            LastName = lastName;
            MiddleName = middleName;
            Titles = titles;
            WorkExperience = workExperience;
            Description = description;
            Photo = photo;
            Stats = stats;
            Informations = informations;
            Specialties = specialties;
        }

        public void AddInformation(Information information)
        {
            if (Informations.All(i => i.Id != information.Id))
            {
                Informations.Add(information);
            }
        }
    }
}

File: src/Web/ClientApp/src/app/components/doctors/doctor.ts
import {IStats, Stats} from "./stats";
import {IInformation, Information} from "./information";
import {ISpecialty, Specialty} from "../specialty/specialty";

export interface IDoctor {
  firstName: string;
  lastName: string;
  middleName: string;
  titles: string;
  workExperience: number;
  description: string;
  photo: string;

  stats: IStats | undefined;
  informations: IInformation[] | [];
  specialties: ISpecialty[] | [];
}

export class Doctor implements IDoctor {
  id: number;
  description: string;
  firstName: string;
  informations: IInformation[] | [];
  lastName: string;
  middleName: string;
  photo: string;
  stats: IStats | undefined;
  specialties: ISpecialty[] | [];
  titles: string;
  workExperience: number;

  static deserialize(data: any): Doctor {
    const doctor = new Doctor()

    doctor.id = data.id;
    doctor.firstName = data.firstName;
    doctor.lastName = data.lastName;
    doctor.middleName = data.middleName;
    doctor.titles = data.titles;
    doctor.workExperience = data.workExperience;
    doctor.description = data.description;
    doctor.photo = data.photo;

    if (data.hasOwnProperty('stats')) {
      doctor.stats = Stats.deserialize(data.stats);
    }

    if (data.hasOwnProperty('informations') && Array.isArray(data.informations)) {
      doctor.informations = data.informations.map(i => Information.deserialize(i));
    }

    if (data.hasOwnProperty('specialties') && Array.isArray(data.specialties)) {
      doctor.specialties = data.specialties.map(i => Specialty.deserialize(i.specialty));
    }

    return doctor;
  }

  public fullName(): string {
    return `${this.lastName} ${this.firstName} ${this.middleName}`
  }
}
